### ISSUE TITLE
New feature: display a list of reverse imports/dependencies

### ISSUE THREAD

**Issue by AARON-CLARK** [pharmaR/riskassessment#261]
Suggested by Andy. AKA - a dependency tree. This will help users understand what packages the package of interest was built on. `riskmetric` contains this information (the list of pkgs).

**Comment by Robert-Krajcik** [comment]
I see we can get useful information out of the `riskmetric` package
```
pkg <- "stringr"
pkg_ref <- riskmetric::pkg_ref(pkg)
deps <- riskmetric::assess_dependencies(pkg_ref)
deps
             package    type
2    glue (>= 1.2.0) Imports
3           magrittr Imports
4 stringi (>= 1.1.7) Imports
```
Next question is, how to display it

**Comment by AARON-CLARK** [comment]
> Next question is, how to display it

That's a really good question @Robert-Krajcik. Maybe we should discuss at our meetup tomorrow.

Off hand, I think that ultimately examining the pkg dependencies its tied to to the pkg maintenance, since a reviewer would want to ensure that pkg X isn't relying on too many pkgs or any pkgs that are obscure or not well supported and make life harder for pkg X in the future. 

On the flip side, if we want to include reverse dependencies (pkgs that rely on pkg X) via `riskmetric::assess_reverse_dependencies(pkg_ref)`, that is kind of a measure of "community usage". 

Or maybe both deps and reverse deps deserve their own tab? We would just have to reduce the horizontal length of these tabs, perhaps by text-wrapping. In general, the final display (wherever it ends up) should separate pkgs into true "dependencies", "imports", "suggests", etc. Note: there are pkgs designed to visually help users understand what pkg functions use which pkg, such as [miniCRAN](https://cran.r-project.org/web/packages/miniCRAN/vignettes/miniCRAN-dependency-graph.html). But I'm not a huge fan of this visual. Ideally, it would be nice to have a button users could click to use `riskmetric` to score each dep / reverse dep too!

**Comment by AARON-CLARK** [comment]
Hmmm, something cool I just found is actually just a URL that visualizes dependencies for you. Perhaps we can display that in an iframe without our shiny app?

Example using `tidyCDISC`: https://depends.rpkg.net/package/tidyCDISC/

Or, this pkg shows promise: https://cran.rstudio.com/web/packages/pkggraph/vignettes/vignette_pkggraph.html

**Comment by Robert-Krajcik** [comment]
This just searches down one level (non-recursive)

```{r, eval=TRUE}
getDependencyTree <- function(pack, i = -1) {
  if(i == -1) cat(pack, "\n")
  i <- i + 1
  packages <- riskmetric::pkg_ref(pack) %>%  riskmetric::assess_dependencies()  
  if (purrr::is_empty(packages)) {
    return() }
  else {packages <- pull(packages, package)
  }
  for(pkg in packages) { 
    for(n in 0:i) {
      cat(" ")
    }
    cat("|", pkg, "\n")
  }
} ## end function getDependencyTree()
getDependencyTree("rvest")
```

rvest 
 | httr (>= 0.5) 
 | lifecycle (>= 1.0.0) 
 | magrittr 
 | rlang (>=
0.4.10) 
 | selectr 
 | tibble 
 | xml2 (>= 1.3)

**Comment by Robert-Krajcik** [comment]
Per Bi-weekly riskasessment meeting 3/28/23
add `dependencies` score (as a card) to `maintenance_metrics` tab
add `reverse_dependencies` score (as a card) to `community_metrics` tab
create new `dependencies` tab

**Comment by Robert-Krajcik** [comment]
I ended up using counts instead of score as I am not sure anyone outside of `{riskmetric}` understands these scores.
I have found that the `reverse_dependencies` card is a better fit with maintenance metrics.
Trying to put it on community usage metrics was like trying to fit a square peg in a round hole.
Here's what I have so far:

- Added dependencies and reverse_dependencies (counts) to maintenance metrics
- Card elements appear in row-major order instead of down the column
- Export_help (often) shows a percentage instead of NA
- Font sizes are all in relation to viewport window (vw)
- Fixed bug in insert_community_metrics_to_db

**Comment by Robert-Krajcik** [comment]
I'm seriously considering the `{deepdep}` package after spending a day wrangling with `[pkggraph}`
https://dominikrafacz.github.io/deepdep/articles/deepdep-comparison.html

- It doesn't require me to initialize anything, like `{pkggraph}` does
- I can feed it dependency (and reverse dependency) data right from `riskmetric::assess_dependencies()`
- It is the only package out there that uses `ggplot2`

**Comment by AARON-CLARK** [comment]
Hey @Robert-Krajcik, I found a unique/ intuitive way to display package dependencies visually. [This app](https://gallery.shinyapps.io/cran-explorer/) is listed in the shinyapps gallery. If you scroll down to the "Explorer" section, you'll see an "ego network" graph for reverse dependencies. If you select a pkg that's used quite heavily, like `dplyr`, you'll get something that's quite hard to understand, but if you pick something in the middle, like `janitor` (which has 99 reverse deps) it shows other packages that use Janitor, and if there's other packages that use those packages, it shows those too! The more pkgs that uses a pkg the bigger the node is.

The code is available here: 
https://github.com/nz-stefan/cran-explorer/blob/master/app/modules/mod_graph_network.R

![Image](https://user-images.githubusercontent.com/17878953/231574288-5208c6c7-fcdf-4b2f-8f29-4e5d182434d6.png)

**Comment by Robert-Krajcik** [comment]
Just a note on displaying dependencies, using the latest version of `{forcats}` -- `1.0.0`
Here is what is in the `DESCRIPTION` file.
```{r, eval=FALSE}
Imports: 
    cli (>= 3.4.0),
    glue,
    lifecycle,
    magrittr,
    rlang (>= 1.0.0),
    tibble
```
There are no DEPENDS or LINKINGTO statements.
This is what `{riskmetric}` sees
```{r, eval=TRUE}
pkg_name <- "forcats"
riskmetric::pkg_ref(pkg_name) %>%  riskmetric::assess_dependencies()
```
This is what `remotes::package_deps()` sees:
```{r, eval=TRUE}
remotes::package_deps(pkg_name, dependencies = c("Depends", "Imports", "LinkingTo")) 
```
Finally, this is what package `{deepdep}` sees:
```{r, eval=TRUE}
deepdep::plot_dependencies(pkg_name, "circular")
```

**Comment by AARON-CLARK** [comment]
closed with #480
